File: typescript/src/resources/oauth.ts
import { createHash, randomBytes } from "node:crypto";
import type { TokenResponse } from "../types.js";
import { EPostakError } from "../utils/errors.js";

/**
 * Helpers for the **integrator-initiated** OAuth `authorization_code` + PKCE
 * flow. Use these from your own backend when you want to onboard an
 * end-user firm into ePošťák from inside your own application — the user
 * clicks a "Connect ePošťák" button in your UI, lands on our `/oauth/authorize`
 * consent page, and we redirect back to your `redirect_uri` with a `code`.
 *
 * This is independent of the regular `EPostak.auth.token({ apiKey })` flow
 * (which uses `client_credentials`). Use one or the other depending on how
 * the firm is linked to you.
 *
 * The OAuth token endpoint lives at `https://epostak.sk/api/oauth/token` —
 * **not** under `/api/v1` — so this resource bypasses `EPostakConfig.baseUrl`.
 *
 * @example
 * ```typescript
 * import { OAuth } from "@epostak/sdk";
 *
 * // 1. On every onboarding attempt, generate a fresh PKCE pair.
 * const { codeVerifier, codeChallenge } = OAuth.generatePkce();
 * await sessions.put(req.session.id, codeVerifier);
 *
 * // 2. Build the authorize URL and redirect the user.
 * const url = OAuth.buildAuthorizeUrl({
 *   clientId: process.env.EPOSTAK_OAUTH_CLIENT_ID!,
 *   redirectUri: "https://your-app.com/oauth/epostak/callback",
 *   scope: "firm:read firm:manage document:send",
 *   state: req.session.id,
 *   codeChallenge,
 * });
 * res.redirect(url);
 *
 * // 3. On callback, exchange the code for a token pair.
 * const tokens = await OAuth.exchangeCode({
 *   code: req.query.code,
 *   codeVerifier: await sessions.get(req.query.state),
 *   clientId: process.env.EPOSTAK_OAUTH_CLIENT_ID!,
 *   clientSecret: process.env.EPOSTAK_OAUTH_CLIENT_SECRET!,
 *   redirectUri: "https://your-app.com/oauth/epostak/callback",
 * });
 * ```
 */
export class OAuth {
  /** Default origin for ePošťák OAuth endpoints. Override for staging. */
  static readonly DEFAULT_ORIGIN = "https://epostak.sk";

  /**
   * Generate a fresh PKCE code-verifier + S256 code-challenge pair.
   *
   * The `codeVerifier` is 43 base64url characters (≈256 bits of entropy).
   * Store it server-side keyed by `state` — you must NOT round-trip it
   * through the user's browser, that defeats PKCE.
   */
  static generatePkce(): { codeVerifier: string; codeChallenge: string } {
    const codeVerifier = randomBytes(32).toString("base64url");
    const codeChallenge = createHash("sha256")
      .update(codeVerifier)
      .digest("base64url");
    return { codeVerifier, codeChallenge };
  }

  /**
   * Build a `/oauth/authorize` URL the integrator can redirect the user to.
   * Uses `response_type=code`, `code_challenge_method=S256`.
   *
   * @param params.scope - Space-separated subset of registered scopes. Omit
   *   to receive the full registered scope list on the consent screen.
   * @param params.origin - Override the host (defaults to ePošťák production).
   */
  static buildAuthorizeUrl(params: {
    clientId: string;
    redirectUri: string;
    codeChallenge: string;
    state: string;
    scope?: string;
    origin?: string;
  }): string {
    const url = new URL(
      "/oauth/authorize",
      params.origin ?? OAuth.DEFAULT_ORIGIN,
    );
    url.searchParams.set("client_id", params.clientId);
    url.searchParams.set("redirect_uri", params.redirectUri);
    url.searchParams.set("response_type", "code");
    url.searchParams.set("code_challenge", params.codeChallenge);
    url.searchParams.set("code_challenge_method", "S256");
    url.searchParams.set("state", params.state);
    if (params.scope) url.searchParams.set("scope", params.scope);
    return url.toString();
  }

  /**
   * Exchange an authorization `code` for a `TokenResponse` on the OAuth
   * token endpoint. Hits `${origin}/api/oauth/token` directly — does not
   * route through `EPostakConfig.baseUrl`, since OAuth lives outside `/api/v1`.
   *
   * The returned access token is a 15-minute JWT; the refresh token is
   * 30-day rotating. Persist both server-side keyed by your firm record.
   */
  static async exchangeCode(params: {
    code: string;
    codeVerifier: string;
    clientId: string;
    clientSecret: string;
    redirectUri: string;
    origin?: string;
  }): Promise<TokenResponse> {
    const url = new URL(
      "/api/oauth/token",
      params.origin ?? OAuth.DEFAULT_ORIGIN,
    );
    const body = new URLSearchParams({
      grant_type: "authorization_code",
      code: params.code,
      code_verifier: params.codeVerifier,
      client_id: params.clientId,
      client_secret: params.clientSecret,
      redirect_uri: params.redirectUri,
    });

    const res = await fetch(url.toString(), {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        Accept: "application/json",
      },
      body: body.toString(),
    });

    const text = await res.text();
    let parsed: unknown = {};
    try {
      parsed = text ? JSON.parse(text) : {};
    } catch {
      parsed = { error: { message: text } };
    }

    if (!res.ok) {
      throw new EPostakError(
        res.status,
        parsed as Record<string, unknown>,
        res.headers,
      );
    }
    return parsed as TokenResponse;
  }
}

File: dotnet/src/EPostak/Resources/OAuth.cs
using System.Net.Http.Headers;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using EPostak.Models;

namespace EPostak.Resources;

/// <summary>
/// Stateless helpers for the <strong>integrator-initiated</strong> OAuth
/// <c>authorization_code</c> + PKCE flow. Use these from your own backend when
/// you want to onboard an end-user firm into ePošťák from inside your own
/// application — the user clicks a "Connect ePošťák" button in your UI, lands
/// on the ePošťák <c>/oauth/authorize</c> consent page, and ePošťák redirects
/// back to your <c>redirect_uri</c> with a <c>code</c>.
/// <para>
/// This is independent of the regular <see cref="AuthResource.TokenAsync"/>
/// flow (which uses <c>client_credentials</c>). Pick one or the other depending
/// on how the firm is linked to you.
/// </para>
/// <para>
/// The OAuth token endpoint lives at <c>https://epostak.sk/api/oauth/token</c> —
/// <strong>not</strong> under <c>/api/v1</c> — so this helper bypasses the
/// configured <see cref="EPostakConfig.BaseUrl"/>.
/// </para>
/// </summary>
/// <example>
/// <code>
/// // 1. On every onboarding attempt, generate a fresh PKCE pair.
/// var (codeVerifier, codeChallenge) = OAuth.GeneratePkce();
/// sessions[req.SessionId] = codeVerifier;
///
/// // 2. Build the authorize URL and redirect the user.
/// var url = OAuth.BuildAuthorizeUrl(
///     clientId: Environment.GetEnvironmentVariable("EPOSTAK_OAUTH_CLIENT_ID")!,
///     redirectUri: "https://your-app.com/oauth/epostak/callback",
///     codeChallenge: codeChallenge,
///     state: req.SessionId,
///     scope: "firm:read firm:manage document:send");
/// return Redirect(url);
///
/// // 3. On callback, exchange the code for a token pair.
/// var tokens = await OAuth.ExchangeCodeAsync(
///     code: req.Query["code"]!,
///     codeVerifier: sessions[req.Query["state"]!],
///     clientId: Environment.GetEnvironmentVariable("EPOSTAK_OAUTH_CLIENT_ID")!,
///     clientSecret: Environment.GetEnvironmentVariable("EPOSTAK_OAUTH_CLIENT_SECRET")!,
///     redirectUri: "https://your-app.com/oauth/epostak/callback");
/// </code>
/// </example>
public static class OAuth
{
    /// <summary>Default origin for ePošťák OAuth endpoints. Override for staging.</summary>
    public const string DefaultOrigin = "https://epostak.sk";

    /// <summary>
    /// Generate a fresh PKCE code-verifier + S256 code-challenge pair.
    /// <para>
    /// The <c>CodeVerifier</c> is 43 base64url characters (≈256 bits of
    /// entropy). Store it server-side keyed by <c>state</c> — you must NOT
    /// round-trip it through the user's browser, that defeats PKCE.
    /// </para>
    /// </summary>
    /// <returns>
    /// Tuple of <c>(CodeVerifier, CodeChallenge)</c> where the challenge is
    /// <c>base64url(SHA256(codeVerifier))</c>.
    /// </returns>
    public static (string CodeVerifier, string CodeChallenge) GeneratePkce()
    {
        var entropy = RandomNumberGenerator.GetBytes(32);
        var codeVerifier = Base64UrlEncode(entropy);
        var hash = SHA256.HashData(Encoding.ASCII.GetBytes(codeVerifier));
        var codeChallenge = Base64UrlEncode(hash);
        return (codeVerifier, codeChallenge);
    }

    /// <summary>
    /// Build a <c>/oauth/authorize</c> URL the integrator can redirect the user
    /// to. Always sets <c>response_type=code</c> and
    /// <c>code_challenge_method=S256</c>.
    /// </summary>
    /// <param name="clientId">Registered OAuth client id.</param>
    /// <param name="redirectUri">Exact-match registered redirect URI.</param>
    /// <param name="codeChallenge">From <see cref="GeneratePkce"/>.</param>
    /// <param name="state">CSRF/session token; echoed back on the callback.</param>
    /// <param name="scope">Optional space-separated scope subset. Omit to receive
    /// the full registered scope list on the consent screen.</param>
    /// <param name="origin">Override the host (defaults to <see cref="DefaultOrigin"/>).</param>
    /// <returns>Absolute authorize URL.</returns>
    public static string BuildAuthorizeUrl(
        string clientId,
        string redirectUri,
        string codeChallenge,
        string state,
        string? scope = null,
        string? origin = null)
    {
        var query = new List<KeyValuePair<string, string>>
        {
            new("client_id", clientId),
            new("redirect_uri", redirectUri),
            new("response_type", "code"),
            new("code_challenge", codeChallenge),
            new("code_challenge_method", "S256"),
            new("state", state),
        };
        if (!string.IsNullOrEmpty(scope))
            query.Add(new KeyValuePair<string, string>("scope", scope));

        var baseUrl = (string.IsNullOrEmpty(origin) ? DefaultOrigin : origin).TrimEnd('/');
        var encoded = string.Join("&", query.Select(kv =>
            $"{Uri.EscapeDataString(kv.Key)}={Uri.EscapeDataString(kv.Value)}"));
        return $"{baseUrl}/oauth/authorize?{encoded}";
    }

    /// <summary>
    /// Exchange an authorization <c>code</c> for an access + refresh token pair
    /// on the OAuth token endpoint. Hits <c>${origin}/api/oauth/token</c>
    /// directly — does not route through <see cref="EPostakConfig.BaseUrl"/>.
    /// <para>
    /// The returned access token is a 15-minute JWT; the refresh token is
    /// 30-day rotating. Persist both server-side keyed by your firm record.
    /// </para>
    /// </summary>
    /// <param name="code">The authorization code from the callback.</param>
    /// <param name="codeVerifier">The verifier paired with the
    /// <c>code_challenge</c> used when starting the flow.</param>
    /// <param name="clientId">Integrator OAuth client id.</param>
    /// <param name="clientSecret">Integrator OAuth client secret.</param>
    /// <param name="redirectUri">Must match the URI used in
    /// <see cref="BuildAuthorizeUrl"/>.</param>
    /// <param name="origin">Override the host (defaults to <see cref="DefaultOrigin"/>).</param>
    /// <param name="httpClient">Optional shared <see cref="HttpClient"/>. When
    /// <c>null</c>, a transient client is created for this call.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Access + refresh token pair.</returns>
    /// <exception cref="EPostakException">On non-2xx response or transport error.</exception>
    public static async Task<TokenResponse> ExchangeCodeAsync(
        string code,
        string codeVerifier,
        string clientId,
        string clientSecret,
        string redirectUri,
        string? origin = null,
        HttpClient? httpClient = null,
        CancellationToken ct = default)
    {
        var baseUrl = (string.IsNullOrEmpty(origin) ? DefaultOrigin : origin).TrimEnd('/');
        var url = $"{baseUrl}/api/oauth/token";

        var form = new Dictionary<string, string>
        {
            ["grant_type"] = "authorization_code",
            ["code"] = code,
            ["code_verifier"] = codeVerifier,
            ["client_id"] = clientId,
            ["client_secret"] = clientSecret,
            ["redirect_uri"] = redirectUri,
        };

        using var request = new HttpRequestMessage(HttpMethod.Post, url)
        {
            Content = new FormUrlEncodedContent(form),
        };
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

        var ownClient = httpClient is null;
        var client = httpClient ?? new HttpClient();
        try
        {
            HttpResponseMessage response;
            try
            {
                response = await client.SendAsync(request, ct).ConfigureAwait(false);
            }
            catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException)
            {
                throw new EPostakException($"Network error: {ex.Message}", ex);
            }

            using (response)
            {
                var payload = await response.Content.ReadAsStringAsync(ct).ConfigureAwait(false);
                if (!response.IsSuccessStatusCode)
                {
                    throw new EPostakException(
                        (int)response.StatusCode,
                        string.IsNullOrEmpty(payload) ? "OAuth token exchange failed" : payload);
                }
                if (string.IsNullOrEmpty(payload))
                {
                    return new TokenResponse();
                }
                return JsonSerializer.Deserialize<TokenResponse>(payload, HttpRequestor.JsonOptions)
                    ?? new TokenResponse();
            }
        }
        finally
        {
            if (ownClient) client.Dispose();
        }
    }

    private static string Base64UrlEncode(byte[] bytes)
    {
        return Convert.ToBase64String(bytes)
            .TrimEnd('=')
            .Replace('+', '-')
            .Replace('/', '_');
    }
}
